src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -7715,34 +7729,107 @@ private ExpressionAst MemberAccessRule(ExpressionAst targetExpr, Token operatorT
                 member = GetSingleCommandArgument(CommandArgumentContext.CommandArgument) ??
                     new ErrorExpressionAst(ExtentOf(targetExpr, operatorToken));
             }
-            else
+            else if (_ungotToken == null)
             {
+                // Member name may be an incomplete token like `$a.$(Command-Name`; we do not look for generic args or
+                // invocation token(s) if the member name token is recognisably incomplete.
+                genericTypeArguments = GenericMethodArgumentsRule(out rBracket);
                 Token lParen = NextInvokeMemberToken();
```

Comment:
I was wrong about this suggestion. It turns out we indeed cannot avoid `resync` because `NextMemberAccessToken()` also assumes `_ungotToken == null`. I have reverted this change in my commit.